aten/src/ATen/native/TensorIndexing.h

```diff
@@ -66,8 +91,8 @@ struct CAFFE2_API TensorIndex final {
   TensorIndex(at::indexing::EllipsisIndexType);
   TensorIndex(const char *str);
 
-  // Case 3: Integer value
-  TensorIndex(int64_t integer);
+  // Case 3: Integer value (the tensor form can optionally be provided)
+  TensorIndex(int64_t integer, Tensor tensor = undefined_tensor);
```

Comment:
I'm pretty unsure what is meant by "tensor form". Is that the case when I do `tensor[tensor2]` index? But I don't understand why would you ram this into the case for integer values...

Reply:
Reading through the old code that produces slices, what I think is going on is you get passed in a Tensor when you are tracing integer arguments through sliced forms. This should be *explicitly* stated as the motivation for all of the Tensor variants.

Reply:
Additionally, there seems to be an invariant that either the tensors are set, or the integers are set, but not both. The way the code is structured today doesn't make this invariant clear; at the very least there should be comments, and even better would be to make the illegal state unrepresentable.

Reply:
For example, it would be a lot clearer if the Tensors were `optional<Tensor>` to show that they aren't necessarily recorded.  But more generally speaking, I'm not sure if it's worth adding trace support to the C++ side, as you can't transparently interchange between an int and a Tensor (unlike in Python). So the tracing logic ought to all be kept at the Python binding side, and now you don't have to goop up the C++ indexing structs with this weird thing.

Reply:
I agreed, just realized that I don't need all the tracing logic in C++ indexing, and can keep them in the Python binding code only.